Feature: OIDC & social authentication

mpflanzer opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

<!--
Please include a summary of the change and which issue is fixed (if any) and any relevant motivation / context. List any dependencies that are required for this change. If appropriate, please include an explanation of how your proposed change can be tested. Screenshots and / or videos can also be helpful if appropriate.
-->

This PR is adding support to authenticate via external OIDC providers (and other "social accounts" probably). Existing users can be linked to social accounts and new accounts are automatically created the first time a login with an unknown social account is done (behaviour can be changed in the settings). See https://github.com/paperless-ngx/paperless-ngx/discussions/4006 for more details.

<!--
⚠️ Important: Pull requests that implement a new feature *should almost always target an existing feature request*. This is in order to balance the work of implementing and maintaining new features vs. community-interest. If that is not currently the case, please open a feature request instead of this PR to gather feedback from both users and the project maintainers.
-->

Closes #4006 

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [ ] Bug fix: non-breaking change which fixes an issue.
- [x] New feature: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [ ] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

mpflanzer committed:
Feature: OIDC support

mpflanzer committed:
Add API for django messages

mpflanzer committed:
Fix exporter tests

mpflanzer commented:
If you prefer to fix the failing exporter tests in a different way I'm open to suggestions

stumpylog commented:
One thing I see missing is documentation.  What are the new settings?  And how does `SOCIALACCOUNT_PROVIDERS` get configured by a user?

I don't know much about this area, so I need that to try and test with Authelia.

shamoon committed:
Fix frontend coverage

shamoon committed:
Discard subscriptions

shamoon committed:
Correct test name typo

shamoon commented on src/paperless/views.py:
```diff
@@ -171,3 +174,70 @@ class ApplicationConfigurationViewSet(ModelViewSet):
+    def post(self, request, *args, **kwargs):
+        user = self.request.user
+
+        user.socialaccount_set.get(pk=request.data["id"]).delete()
+
+        return Response(
+            request.data["id"],
```
The security warning seems valid, also will this validate the account? Perhaps something like:

```suggestion
        account = user.socialaccount_set.get(pk=request.data["id"])
        if account is not None:
            account.delete()
            return Response(
                account.id,
            )
        else:
            return HttpResponseBadRequest(
                "Social account not found",
```

mpflanzer replied:
Isn't get throwing an exception if it cannot find exactly one object? Than my version be save but indeed doesn't return the best error response. How about:

```
try:
   account = user.socialaccount_set.get(pk=request.data["id"])
   account.delete()
   return Response(account.id)
except SocialAccount.DoesNotExist:
   return HttpResponseBadRequest("Social account not found")
```

shamoon replied:
Sure, same idea I think

shamoon commented:
@mpflanzer Did you intentionally disable pushing to your branch or not intentional? Theres a checkbox somewhere I think. If you would prefer we not do that its fine, but I have a few commits that fix the frontend coverage and other small frontend-only things

shamoon committed:
Formatting

shamoon commented on src/paperless/templates/account/login.html:
Why are there now two of these templates? E.g. vs src/documents/templates/registration/login.html

Also this now throws an error for me because of the password_reset links

mpflanzer replied:
Should be fixed now. I removed the old login template. And I forgot about the password reset as I hadn't configured the email server.

shamoon commented on src-ui/src/app/components/common/profile-edit-dialog/profile-edit-dialog.component.html:
UI here needs some work. I'll need to get a working example, but the screenshot is a little awkward, the text doesnt really make sense either:

<img width="573" alt="Screenshot 2023-12-31 at 11 24 53 AM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/ddcea3d0-961f-4ed3-8224-00691c9602f7">

shamoon reviewed (commented):
Yea docs will be helpful, I quickly tried to set this up with e.g. the code below and no dice. I can add the provider via the Django admin, but I had sorta thought thats not necessary?

```
PAPERLESS_APPS="allauth.socialaccount.providers.github"
SOCIALACCOUNT_PROVIDERS={
    "openid_connect": {
        "APPS": [
            {
                "provider_id": "github",
                "name": "Github",
                "client_id": "*",
                "secret": "*",
                "settings": {
                },
            },
        ],
    },
    'github': {
        'SCOPE': [
            'user',
            'repo',
            'read:org',
        ],
    }
}
```

shamoon committed:
Update profile UI for social accounts

shamoon commented:
Another big question is what happens when a new user gets created. I havent tested that, but I presume they will have 0 permissions to start, which is an issue.

Making progress on the UI, again let us know if youre OK with pushing to your branch.

<img width="634" alt="Screenshot 2023-12-31 at 12 09 30 PM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/3b86ac9e-c4ac-4d65-9dc2-e4081bcc140b">

mpflanzer commented:
> @mpflanzer Did you intentionally disable pushing to your branch or not intentional? Theres a checkbox somewhere I think. If you would prefer we not do that its fine, but I have a few commits that fix the frontend coverage and other small frontend-only things

No not intentionally at all. Changes are welcome. It seems that option doesn't exist because I created the PR based on a fork owned by an organization and not my personal account. :see_no_evil: Do you want me to close this one and open a new one from my personal account?


> One thing I see missing is documentation. What are the new settings? And how does `SOCIALACCOUNT_PROVIDERS` get configured by a user?
> 
> I don't know much about this area, so I need that to try and test with Authelia.

Here is a pretty comprehensive list of how to setup providers: https://docs.allauth.org/en/latest/socialaccount/providers/index.html Authelia is not on that list but I presume it's similar to Keycloak (which I used for testing).

shamoon commented:
> It seems that option doesn't exist because I created the PR based on a fork owned by an organization and not my personal account.

Oh, interesting. What if you give us access to the repository (not the org)? I think that would do it?

mpflanzer commented:
Done. Let me know if I should add more people.

mpflanzer commented on src-ui/src/app/components/common/profile-edit-dialog/profile-edit-dialog.component.spec.ts:
```diff
@@ -265,28 +265,26 @@ describe('ProfileEditDialogComponent', () => {
+    const errorSpy = jest.spyOn(toastService, 'showError')
 
-    expect(getSpy).toHaveBeenCalled()
     expect(component.socialAccounts).toContainEqual(socialAccount)
```
Is it guaranteed that the profile observable in `ngOnInit` has populated the `socialAccounts` member already at this point? I thought some synchronization would be needed.

shamoon replied:
For the purposes of this test, it’s fine I think

mpflanzer commented:
> Another big question is what happens when a new user gets created. I havent tested that, but I presume they will have 0 permissions to start, which is an issue.

Correct, but we can override the `populate_user` method (https://github.com/pennersr/django-allauth/blob/4c3b84a058b4bf3f9ea8fac72e053e90039bbebc/allauth/socialaccount/adapter.py#L105) in the social account adapter to customize what is supposed to happen.

mpflanzer committed:
Remove superseded login form

mpflanzer committed:
Fix password reset

mpflanzer committed:
Move all remaining templates

shamoon commented on src-ui/src/app/components/app-frame/app-frame.component.ts:
```diff
@@ -92,6 +94,25 @@ export class AppFrameComponent
     this.tasksService.reload()
+
+    this.messagesService
+      .get()
```
Can you explain exactly why this is needed (I get what it does I mean what situation it’s needed)? This service now runs one time only, which seems like a bit of a waste

mpflanzer replied:
In some cases allauth is just indicating an error via a django message. For example if the account cannot be linked: https://github.com/pennersr/django-allauth/blob/4c3b84a058b4bf3f9ea8fac72e053e90039bbebc/allauth/socialaccount/helpers.py#L197

I just went for the minimal implementation to get access to the messages. And in this specific case for authentication I think it always enough to just check on page load of the frontend for new messages. I guess anything else would need to be based on polling for messages from django. I'm not aware of any hook we could use to have django automatically push new messages.

Or if you have another idea for the error handling I'm also open to that.

shamoon replied:
So if I understand correctly, if you want to use these new buttons under profile to login and theres an error:
1. User clicks button from frontend --> opens login URL in new tab
2. Proceeds via login flow in whatever app
3. User is redirected back to paperless in the same browser window, at which point app loads & checks for Django messages that might be present.

In this case, the login button probably shouldn't have `_blank` since you end up with two windows of the app, I'll change that.

For the error messages, I guess we dont have much control over what happens once the user enters the allauth flow process. Maybe we can embed the message in the page or something simple like that, rather than having a whole new API endpoint, frontend service, etc. that only gets used once. 

But yea, I need to think about it / play with it.

mpflanzer replied:
I just found that all calls to `add_message` from `allauth` are going through the adapter which we subclass anyway already: https://github.com/pennersr/django-allauth/blob/4c3b84a058b4bf3f9ea8fac72e053e90039bbebc/allauth/account/adapter.py#L367

So we could hook into that to transport the messages more like push notifications to the frontend. Could the websocket be reused for this?

shamoon replied:
Yea I thought about the websocket. It would need to be adapted quite a bit as at the moment it's only expecting consumer messages, which also seems sorta not worth it because this basically only happens once, right? But the other question is whether the timing would work, i.e. when the backend calls `add_message` the user is still in the process of being redirected in which case by the time they load the frontend UI that message wont be received (I think). Django queues them until someone loads a page which is why it works there, your current implementation effectively does the same i.e. just pulls from the queue once the UI loads.

Will keep brainstorming, I havent looked at it yet much myself. Its not the end of the world if it stays like this.

shamoon replied:
Ok, I just pushed a change that should sort this, I think. Any pending messages are just embedded in the html and the frontend picks them up on start, once. I tested this out a bit, of course let me know if you notice any issues.

<img width="402" alt="Screenshot 2024-01-05 at 10 34 33 AM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/f9d63e65-edf4-4385-bcdf-dfa69b4259cd">

mpflanzer replied:
Looks good to me

shamoon commented on src/paperless/templates/socialaccount/login.html:
Do these templates all need to be in the paperless module as opposed to leaving them under documents?

mpflanzer replied:
Hmm, probably they can be moved over to documents again. At the beginning I was a bit confused because I expected them to be in the main paperless app. Since they're not about the documents at all.
If you prefer I can give it a try if they can be moved back.

shamoon replied:
I think in some ways it'd be less 'disruptive' to leave them under documents, yea.

I agree it's a little confusing, 'documents' is basically the module for all of paperless-ngx, 'paperless' is the old original app. At this point the separation is pointless, if I could snap my fingers and merge them I would 😉

mpflanzer replied:
Seems to work just fine to have the templates under `documents`. :+1:

shamoon replied:
cool thanks

shamoon committed:
Add noopener noreferrer to social login button

shamoon committed:
Test doesnt need to be async

mpflanzer committed:
Fix server-side XSS security warning

mpflanzer committed:
Move templates back to documents

shamoon committed:
Dont open social login links in new window

shamoon commented on src/paperless/adapter.py:
Flagging this: we should implement `populate_user ` with some default perms (question of which ones remains), see https://github.com/paperless-ngx/paperless-ngx/pull/5190#issuecomment-1873239956

shamoon replied:
Actually, besides the docs and the backend review, this is the one issue I think needs to be resolved.

> Permissions seems to be restrictive enough at the moment

Please correct me if Im wrong, but I think from the user perspective, there are two cases:

1. User exists and is linked to a social account. In this case the user will by necessity already have sufficient paperless permissions. Right?
2. User is created by this flow. In this case, are we saying the user should have 0 permissions and then have to request them (by what mechanism?)? I think it would make more sense if, in this situation where the admin has explicitly allowed social signups, the user should at least be given some basic permissions. Or ideally a setting for this. Otherwise, I think in this case the user will just come back to the app and basically not be able to do anything. See below...

<img width="1726" alt="Screenshot 2024-02-05 at 11 10 22 AM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/62b11e3e-a355-42b4-aa51-04f83b46d8a8">

mpflanzer replied:
Yes, case 1 should be clear. All permissions remain the same.

For case 2 I don't see a big problem. At least as SW developer it's rather common that you can create an account somewhere but have no permissions. Then you write to an "admin" (via any means you like) and they manually grant you permissions. Ok, not being able to the profile etc. is a bit extreme but wouldn't be a blocker for me either.
We can also add a basic set of settings around managing the profile as default. I think that doesn't cause any harm.

shamoon replied:
I guess this isnt dissimilar to creating a new user in the app in general right now, so we can leave this as-is, perhaps a future PR could implement some kind of default permissions in general (and here).

I added a TODO for the future.

shamoon committed:
Pass Django messages to frontend without API endpoint

shamoon committed:
Refactor to django-messages.service

shamoon committed:
Missed file in last commit

mpflanzer committed:
Add missing template in case of auth error

florianschroen commented:
today i tested the current commit 7ab60027 from the branch mpflanzer-allauth from https://github.com/chickadee-engineering:mpflanzer-allauth.

I got a working login with a OpenID Connect auth done with goAuthentik. The setup may be reproducable with other OIDC solutions like Authelia as well. 

Steps to reproduce:

- Create Authentication Provider of type "OAuth2/OpenID Provider" in goAuthentik
- Copy information from from Provider created provider to notepad:
  - OpenID Configuration Issuer (URL)
  - Client ID
  - Client Secret
- Create Application in goAuthentik
  - UI Settings > Launch URL: "http://localhost:8010" (if you are testing on localhost or the url of your paperless instance)
- switch to paperless-ngx
- goto settings > open Django-Adminpanel > Accounts > Social Apps > add
  - provider: OpenID Connect
  - provider id: choose something which is uniq and describes the current social app, e.g. "auth-mycompany" (is inserted to the login url and referenced in the user accounts)
  - name: e.g. "mycompany" or "auth.mycompany.tld" (is shown as label at the login button)
  - client-id: paste "Client ID" value from notepad
  - secret key: paste "Client Secret" value from notepad
  - settings: {"server_url": "https://..."} - paste your OpenID Configuration Issuer URL from notepad

mpflanzer commented:
> Yea docs will be helpful, I quickly tried to set this up with e.g. the code below and no dice. I can add the provider via the Django admin, but I had sorta thought thats not necessary?

I tried it with Github myself now and got it working with the following config in the `settings.py`:

```
SOCIALACCOUNT_PROVIDERS = {
    "github": {
        "APPS": [
            {
                "provider_id": "github",
                "name": "Github",
                "client_id": "XXXXXX",
                "secret": "XXXXXX",
                }
            ]
    },
}
```
`provider_id` and `name` can be freely chosen. Additionally `allauth.socialaccount.providers.github` needs to be added to the installed apps.

shamoon commented:
> I tried it with Github myself now and got it working with the following config in the `settings.py`:

Yea I used the same code etc. but the thing I was a bit confused / disappointed by was that I still needed to go into the Django admin and set it up, I wasnt clear whether the expectation is that either you set it up via the admin or with these settings or in fact you need both

mpflanzer commented:
> > I tried it with Github myself now and got it working with the following config in the `settings.py`:
> 
> Yea I used the same code etc. but the thing I was a bit confused / disappointed by was that I still needed to go into the Django admin and set it up, I wasnt clear whether the expectation is that either you set it up via the admin or with these settings or in fact you need both

The code snippet you posted was a bit different. The key of the provider needs to be `github` (not `openid_connect`). Basically you need to merge the to sections you had.
I didn't use Django admin for my test. Everything was working by just using the settings file.

shamoon commented:
Ah sweet, thats great. Yea was hoping it would be just set it once kinda thing for users. As I said, when we get there documentation will be good 😆. Presumably a lot of it I'll be linking to all-auth, etc.

Ps. In general, I dont think this PR is very far, but the other thing is we've just released a bunch of new big-ish features so we're just a bit focused on shoring those up for now. But Im hopeful this can eventually get 'over the line', thanks for your patience

florianschroen commented:
I have setup oidc via Django admin and I wonder if modifying settings.py needs to be mention for docker via env variables and how. Because theoretical there can be as many social logins as the admin wants.

shamoon commented on src/paperless/settings.py:
```diff
@@ -409,8 +415,38 @@ def _parse_beat_schedule() -> dict:
+    "allauth.account.auth_backends.AuthenticationBackend",
 ]
 
+SOCIALACCOUNT_PROVIDERS = {
+    # "openid_connect": {
+    #    "APPS": [
+    #        {
+    #            "provider_id": "keycloak",
+    #            "name": "Keycloak",
+    #            "client_id": "paperless",
+    #            "secret": "<CLIENT_SECRET>",
+    #            "settings": {
+    #                "server_url": "https://<KEYCLOAK_SERVER>/realms/<REALM>/.well-known/openid-configuration",
+    #            },
+    #        },
+    #    ],
+    # },
+}
```
```suggestion
SOCIALACCOUNT_PROVIDERS = json.loads(os.getenv(
    "PAPERLESS_SOCIALACCOUNT_PROVIDERS",
)) if "PAPERLESS_SOCIALACCOUNT_PROVIDERS" in os.environ else None
```

Presumably should be something like this so users aren't editing settings.py directly, then you can use e.g.

`PAPERLESS_SOCIALACCOUNT_PROVIDERS='{"github": {"APPS": [{"provider_id": "github","name": "Github New","client_id": "xxx","secret": "xxx"}]}}'` in paperless.conf, docker ENV, etc.

mpflanzer replied:
Sounds good to me. I pushed a change to load the providers from the environment. Do you want a dedicated variable for the provider apps? Or is using the existing `PAPERLESS_APPS` good enough?

shamoon replied:
Cool. I think its fine as-is, tho again I’ll defer to the backend review

mpflanzer committed:
Load social account providers from env

acul009 commented:
I'm very interested in OIDC too.
Can I help in any way?

I have tandoor with django-allauth connected to my authentik instance and I could share the configuration if that would help.

In case the allauth solution isn't welcome, I could also help with implementing a more basic OIDC integration.
I already contributed a (janky) OIDC connector for another open source project. It's not perfect, but if you only support a subset of possible configurations, you can allready connect most OIDC implementations.

Either way, I'd be happy to support this feature :)

mpflanzer commented:
> I'll have to look into this further (I haven't had a chance to deploy paperless yet), but being able to map permissions from a scope/claim would be great

Isn't this kind of the wrong way round? Paperless would ask for a certain scope and the user can decide to grant permission or decline. But for the permissions it shouldn't be the user deciding but some kind of admin. Or is there also a way to configure the OAuth App, for instance in Github, to only allow certain scopes?

0xC0ncord commented:
> Isn't this kind of the wrong way round? Paperless would ask for a certain scope and the user can decide to grant permission or decline. But for the permissions it shouldn't be the user deciding but some kind of admin. Or is there also a way to configure the OAuth App, for instance in Github, to only allow certain scopes?

As I understand it the proper way to do this is actually to associate the user with specific groups in the SSO application and have those groups mapped to groups in Paperless-ngx, then those groups will have the granted permissions associated as configured by the admin.

mpflanzer commented:
But this requires that I have control over the identity provider. If I connect to GitHub for instance I guess there is no way of creating groups etc. So in the first case that would be nice to leverage but it wouldn't solve the permission question in general.

But would it be an issue to start with no permissions and wait for an admin to assign permissions? Currently "signup" is also manually done by an admin, right?

johnyb0y commented:
There are several ways to handle this from an identity provider perspective. In for example AzureAD (or "EntraID" as it is now called) you can work both with groups to assign certain permissions to certain people, and also often you can assign a profile to a person, which is basically an additional information which gets sent to the attached application and then handled in a certain way.

As others have said, this is dependent on both the respective identity provider and how the application has implemented its rights management and there is not always a uniform way.

As a first step, just the possibility to assign a certain default profile to people logging in via the identity provider for the first time would be a good starting point in my opinion.

TheRealGramdalf commented:
> Isn't this kind of the wrong way round? Paperless would ask for a certain scope and the user can decide to grant permission or decline. But for the permissions it shouldn't be the user deciding but some kind of admin. Or is there also a way to configure the OAuth App, for instance in Github, to only allow certain scopes?

Yes, it is typically up to the identity provider. For things like github, where you don't have control, it doesn't make sense - but for something self hosted such as Authelia, Kanidm, or Authentik, you have the ability to map custom claims/scopes. 

In OIDC, a claim is a piece of information about the user. By default, there are standard claims like email address, subject (a UUID for the user), first/last name, preferred name (username/displayname) etc. This can also include information such as membership of a group, which is typically used to indicate permissions.
A scope, on the other hand, is a collection of claims - the default scope `profile` includes many of the things above like email, username etc.

An Oauth2 resource server (Paperless) can request arbitrary scopes and/or claims from the identity provider, which it can use in any way it sees fit. It could take the presence of the scope `paperless-admin` to indicate that a user has administrative access to the paperless server (even if the scope has no claims, more info in a sec). Though I haven't had a chance to deploy paperless yet, I believe I saw that it already has the concept of role-based access control (permissions through groups) - the simplest way to implement this would be to allow the user to map a scope/claim from the IdP to a group in paperless, but it would be better UX wise to have a couple of pre-set scopes/claims (i.e. `paperless-admin`) that automatically map to the relevant group/role in paperless.

One other thing to keep in mind is usernames - some smaller projects incorrectly use `preferred_username` or `username` to associate a user from the IdP to a user in the resource server. This is bad because the uniqueness of these claims is not guaranteed - a user can change these at any time, according to the RFC. Some IdPs do guarantee the uniqueness of these claims, but again, it is not guaranteed according to the RFC. Instead, the claim `sub` should be used. This is typically a non-human readable UUID, but is guaranteed to be immutable per account creation. It is up to the resource server to implement a display name/pretty name, so that the user sees `TheRealGramdalf` in the UI  as opposed to `some-random-uuid`. The claim `preferred_username` is typically used for the display name.

One other thing to note is that with OIDC, the user typically doesn't enter their username/password in the resource servers login UI - instead, they click on a link that takes them to their IdP (such as google, github, etc), where they login with the typical flow. Unless paperless decides to do things differently, there shouldn't be a need to have any correlation between username and UUID in the login page - this would be a purely visual substitution.

Let me know if that makes sense, I can provide further info if need be.

TheRealGramdalf commented:
Also forgot to mention - mapping a scope/claim to permissions in paperless should be off by default, and the user can enable it if needed. By default, users authenticating via OIDC should have no access - they should be able to log in, but not have access to any data, read or write. It is assumed that the IdP is trusted - if the user has a scope/claim (due to the resource server-specific nature of OIDC/Oauth2), it is assumed that the IdP is making sure that a user only has the scopes they should.

Grafana has a very good (and RFC compliant) implementation of OIDC/Oauth2 - here's a [link to the overview](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/generic-oauth/), an [integration planning brief](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/planning-iam-strategy/), and the [authentication configuration reference](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#auth) (and the [oauth2 specific section](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#authgeneric_oauth) of said configuration reference)

mpflanzer commented:
Thanks for the details @TheRealGramdalf. That's matching how I pictured it. `django-allauth` is using the `sub` claim to link a social account to a local user account. And the login happens, as you said, by redirecting to the IdP.

So to summarize a bit the discussion:
- By default users are able to signup via social auth but have no permissions until an admin assigns them some
- Existing local users can link their social account and keep all the permissions they had already
- For convenience it would be nice to derive permissions/groups from claims/scopes returned by the IdP.

The first two points is the current state of the branch. I would propose to keep the third point as a next step out of scope of this PR.

TheRealGramdalf commented:
> The first two points is the current state of the branch. I would propose to keep the third point as a next step out of scope of this PR.

I concur. Is there a docker image I can use to test drive this with my chosen SSO solution, Kanidm? Or would I have to build it from the branch over at https://github.com/chickadee-engineering:mpflanzer-allauth?

florianschroen commented:
You need to build by yourself.

`docker compose -f path/to/choosen-docker-compose.yml build`

to build it. And then start it

`docker compose -f path/to/choosen-docker-compose.yml up -d`

TheRealGramdalf commented:
I was also able to get my provider (Kanidm) functional with the following configuration:
*Replace text in `<brackets>` with the correct value for your case, e.g. `kanidm group create <login-group-name>` becomes `kanidm group create paperless-ngx-openid`*

### Docker image
- Clone the repository: `git clone -b mpflanzer-allauth https://github.com/chickadee-engineering/paperless-ngx`
    - *Using commit `git log`: `commit 9a71c979da873e9b4fe7f98205212049c3f66a0d`*
- Build the docker image: `cd paperless-ngx && docker build --tag paperless-ngx:mpflanzer-allauth --network=host .`

### Kanidm
- Add an Oauth2 resource server  (`kanidm system oauth2 create <client-id> '<Display Name>' 'https://paperless.my.domain.tld'`)
    - Create a group for users that are allowed to log in (`kanidm group create <login-group-name>`)
    - Add scope map for `openid`, `profile`, and `email` (all of which are requested by Paperless-ngx) (`kanidm system oauth2 create-scope-map <client-id> <login-group-name> openid profile email`)
    - Add a user to the group (`kanidm group add-members <login-group-name> <allowed-user-1> <allowed-user-2>`)
- Get the basic secret (`kanidm system oauth2 show-basic-secret <client-id>`)

### Paperless
- Since paperless is behind a proxy, set `PAPERLESS_PROXY_SSL_HEADER='["HTTP_X_FORWARDED_PROTO", "https"]'`
    - *This makes sure paperless is aware that it's running via https, see the [paperless docs](https://docs.paperless-ngx.com/configuration/#PAPERLESS_PROXY_SSL_HEADER) and the [django docs](https://docs.djangoproject.com/en/4.2/ref/settings/#secure-proxy-ssl-header)*
- Enable the generic OpenID provider by setting `PAPERLESS_APPS='allauth.socialaccount.providers.openid_connect'`
    - *For more well known providers, you could use their preset instead*
- Configure the provider via an environment variable: `PAPERLESS_SOCIALACCOUNT_PROVIDERS='{ "openid_connect": { "APPS": [ { "provider_id": "kanidm", "name": "Paperless Display Name", "client_id": "<client-id>", "secret": "alskdjflkassdjf-client-secret-herelakjdsflkjassdf", "settings": { "server_url": "https://auth.domain.tld/oauth2/openid/<client-id>" } } ], "OAUTH_PKCE_ENABLED": "True" } }'`

Readable version:
```jsonc
{
    "openid_connect": {
        "APPS": [
            {
                "provider_id": "kanidm", // Choose whatever ID you want (inserted to the login url and referenced in the user accounts)
                "name": "Paperless Display Name", // Choose a human-readable user facing name (shown as label at the login button)
                "client_id": "<client-id>", // Use the client id from Kanidm
                "secret": "alskdjflkassdjf-client-secret-herelakjdsflkjassdf", // Insert your actual secret here
                "settings": {
                    "server_url": "https://auth.domain.tld/oauth2/openid/<client-id>" // Replace <client-id> with the client id from Kanidm 
                }
            }
        ],
        "OAUTH_PKCE_ENABLED": "True" // Required specifically for Kanidm, since PKCE is enforced by default (not using PKCE is considered insecure)
    }
}
```
>

stumpylog commented:
I have made some progress testing with Authelia as a OIDC provider, but this still needs:

- Documentation.  At a very least, the new settings and ideally some example configurations for the more common providers someone self hosting will use (Authentik and Authelia come to mind).  These could be maintained in the wiki or upstreamed if the provider has that sort of thing in documentation
- A user should be able to set `ACCOUNT_DEFAULT_HTTP_PROTOCOL`, and this should default to `https`, not allauth's default of http.  Secure by default seems best
- I'm fine with initial permissions needing to be added manually by an admin, in this initial support

TheRealGramdalf commented:
As far as initial permissions go, there were a couple things I had in mind:
- It looks like when a user first signs in, they can still upload documents (and therefore use space etc) - that could be a concern especially with things like github, where you can't necessarily control who can access it (as far as I know, I haven't tried yet though)
- Not sure how this would be implemented, but it would be nice if I didn't have to enable OpenID connect with `PAPERLESS_APPS='allauth.socialaccount.providers.openid_connect'`, since I already specified a provider with `PAPERLESS_SOCIALACCOUNT_PROVIDERS`. 
- When I was doing setup I ran across a note saying that documents without an owner are by default visible to everyone - that might also be worth noting, but I'm not quite sure how that would work with the consume folder - if someone uploads documents (i.e. via a generic SMB share), who should be assigned as the owner? Unless user-authenticated uploads are enforced, I don't see an easy way to make this secure by default (apart from disabling the consume folder entirely).
    - The easiest way I can think to solve this long term would be to have a tighter consume integration (keeping in mind that I haven't really _used_ paperless-ngx yet - this could well be implemented already) such that users could connect via WebDAV, SFTP (aka SSHFS) or something else that's bound to their account so documents can be properly assigned. Of course, an admin could create this "integration" with custom tagging rules (via folder) and separate shares, but having it integrated directly would be very nice. 

But I digress - the last point is definitely out of scope of this initial PR - if there isn't an easier solution, I would probably just disable the consume folder by default (making it opt-in for users who enable OIDC) and show the user a warning with the security implications.

shamoon commented:
@mpflanzer just checking in, theres no rush but with 2.4 out we've sorta cleared the way to circle back to this, let us know if youre able to hop back on it.

mpflanzer commented:
@shamoon I'm on vacation next week but might have a bit of time to look into things. I'd like to finish this rather sooner than later.

What are the critical points to be addressed until the first version could be merged?

shamoon commented:
> What are the critical points to be addressed until the first version could be merged?

Sorry I missed your question in the post. If other maintainers have more to add to this list please chime in but IMHO, in no particular order:

- Documentation
- Testing coverage
- Permissions
- Merge conflicts

Enjoy your vacation, FOSS should never supersede that (unless it helps you relax 😜)

compgeniuses commented:
this is sweet

mpflanzer commented:
I didn't find any time last week, and the next weeks also don't look to great. But I had good vacation at least. :grin: Anyway, I'd really like to get this finished soon-ish before it's getting to stale.

- I'll tackle the documentation last as this should be pretty straightforward. But if someone else feels motivated contributions are welcome. :wink: 
- I've rebased it onto the current `dev` branch. The Django translations file still needs to be regenerated now. But I'll wait until the end with this as well.
- I'll look at the test coverage next as the permissions actually seems to be ok for now, i.e. after signup via social account login a user cannot see and to anything

> A user should be able to set ACCOUNT_DEFAULT_HTTP_PROTOCOL, and this should default to https, not allauth's default of http. Secure by default seems best

Done :+1:

Thanks for the additional tests and notes, @TheRealGramdalf 

> It looks like when a user first signs in, they can still upload documents (and therefore use space etc) - that could be a concern especially with things like github, where you can't necessarily control who can access it (as far as I know, I haven't tried yet though)

I tried with an account created by social account login but I don't have the widget to upload documents in the dashboard with that user. Unless I overlooked something that seems to be ok.

> When I was doing setup I ran across a note saying that documents without an owner are by default visible to everyone

I saw that note as well now. But can't confirm that this is indeed the case. I uploaded a document with a regular user and removed the owner afterwards. The social account user cannot access the document. The only thing this user seems to be allowed to see are the statistics in the top right corner, i.e. "Total documents" and "Total characters". Probably not ideal but also not a disaster I would say. Maybe this can also be restricted rather easily. Would be great if someone else could try this as well and confirm that no other information is leaked.

> Not sure how this would be implemented, but it would be nice if I didn't have to enable OpenID connect with PAPERLESS_APPS='allauth.socialaccount.providers.openid_connect', since I already specified a provider with PAPERLESS_SOCIALACCOUNT_PROVIDERS.

I have the feeling this could turn into a rather ugly thing if it's supposed to be somewhat generic for other providers as well. Not sure if I would try to parse the providers to extract the apps from them. Especially since the manual specification is not too much of a hassle either.

mpflanzer committed:
Make allauth generate https links by default

mpflanzer committed:
Fix tests that require the openid-connect app to be installed

mpflanzer commented:
The tests currently fail because they rely on the `openid_connect` provider which is no longer part of the installed apps by default. My attempt to just add it to the `INSTALLED_APPS` for the test (https://github.com/paperless-ngx/paperless-ngx/pull/5190/commits/5b4f47124fd618fd16d9a4150d5bfbc354630842) is not working either.

Any suggestions how to continue from here short of adding it to the `settings.py`?

TheRealGramdalf commented:
Yep - you need to set `PAPERLESS_APPS='allauth.socialaccount.providers.openid_connect`. It's part of my comment above if you need more context (but its easy to skim over at a glance).
If you want to test with a specific provider (instead of a generic one) I believe you just need to replace `openid_connect` with a valid django-allauth provider - I could be wrong on the exact syntax though, I forget exactly where I found that information (I'm on mobile atm)

mpflanzer commented:
I was talking about the automated Python tests not how to set it up manually. Apologies if that wasn't clear. I could of course add the provider used in the tests to the settings/config but since that's the same used for deployment it would mean that this provider app is installed by default. (It's only the about the installed app, the provider config is done in the test)

shamoon commented:
I’m not the best person for backend code review but I’m sure that last commit will not be acceptable. Did you try https://docs.djangoproject.com/en/5.0/topics/testing/tools/#django.test.override_settings ?

TheRealGramdalf commented:
On another note, I would be willing to contribute to the docs if you like - I don't have enough time to polish everything and do the actual PR, but I can certainly write out some info in a similar style to the other paperless-ngx docs, since I'm already quite familiar with deploying OIDC.

shamoon commented:
It’s up to @mpflanzer but that would be fine with us of course. Currently however there is no documentation whatsoever.

mpflanzer commented:
> On another note, I would be willing to contribute to the docs if you like

That would be really nice, thanks! I can polish it in the end, put it in the right place, etc. Already if you provide some content that would help a lot. I granted to write permission on the PR in case you want to push something. Otherwise if you drop comments here I can also integrate it in to the PR.

TheRealGramdalf commented:
Sounds good! Just one thing then - is everything pretty much nailed down at this point or is there still potential for things to change? I can start whenever, but there's not much point writing docs on something that may not be included in the final merge ;)

mpflanzer commented:
> I’m not the best person for backend code review but I’m sure that last commit will not be acceptable. Did you try https://docs.djangoproject.com/en/5.0/topics/testing/tools/#django.test.override_settings ?

Yes, I tried that first (https://github.com/chickadee-engineering/paperless-ngx/commit/5b4f47124fd618fd16d9a4150d5bfbc354630842). But allauth still complained that the provider is not known:

```
FAILED documents/tests/test_api_profile.py::TestApiProfile::test_get_profile - django.core.exceptions.ImproperlyConfigured: unknown provider: openid_connect
```

The additional settings file is not optimal but I guess the two other choices are: remove the test or have the openid_connect provider in the settings file.

mpflanzer commented:
> Sounds good! Just one thing then - is everything pretty much nailed down at this point or is there still potential for things to change?

I don't plan to change anything for this PR. Permissions seems to be restrictive enough at the moment and the signup/login/disconnect flow is fully working. Test coverage passed as well I think on the last run. Let's see what the review brings but I don't expect anything that would change the features.

@shamoon what is the process to initiate the review? Do you want the documentation first as well?

shamoon commented:
> Yes, I tried that first ([chickadee-engineering@5b4f471](https://github.com/chickadee-engineering/paperless-ngx/commit/5b4f47124fd618fd16d9a4150d5bfbc354630842)). But allauth still complained that the provider is not known:

Well, [override_settings](https://docs.djangoproject.com/en/5.0/topics/testing/tools/#django.test.override_settings) not modify, and I would think something like `@override_settings(PAPERLESS_APPS="allauth.socialaccount.providers.openid_connect")` since thats how it will actually be used. But again, Im not the expert here. But removing the tests would not be a solution either, no.

Once the above is sorted I think this can undergo a backend code review.

> Do you want the documentation first as well?

Yes. A backend code review isnt dependent on this, but the PR overall would be.

shamoon committed:
Revert "Fix tests that require the openid-connect app to be installed"

This reverts commit 1e572dc74addc9f4bf990c678849c06252f137dc.

mpflanzer commented:
I tried `override_settings` as well but same result. I guess for the apps Django is doing something extra when it loads the settings which is then not replicated when they are modified via the decorators.

If you think the additional settings file for testing has no chance in the review I'll add the openid_connect app to the real settings because I'm out of ideas how else I could get the test working.

> Yes. A backend code review isnt dependent on this, but the PR overall would be.

PR for sure, yes. I would propose to then start with the review and do the documentation in parallel.

shamoon committed:
Just add oidc to installed apps

shamoon commented:
> If you think the additional settings file for testing has no chance in the review I'll add the openid_connect app to the real settings because I'm out of ideas how else I could get the test working.

Funny, I was just about to write a comment to this effect. I think that'd be fine and is the cleanest way. If anyone disagrees they can chime in during the review.

compgeniuses commented:
i am currently following up on 3 opensource projects that are implementing SSO in one form or another, and its interesting see how the entire process is unfolding. truly openosurce maintenance is not a walk in the park

shamoon committed:
Fix social_account_providers endpoint test

shamoon committed:
Add coverage for adapter

shamoon committed:
Fix local vs ci social_account_providers test

shamoon commented:
In the interest of helping things along I reverted that commit, moved the oidc app into settings and fixed / added some backend tests. LMK thoughts.

mpflanzer committed:
Add comment why openid_connect is installed by default

mpflanzer commented:
Thanks! Looks good to me. I just added a comment to explain why this app is installed by default.

shamoon commented on src/paperless/views.py:
```diff
@@ -168,3 +172,57 @@ class ApplicationConfigurationViewSet(ModelViewSet):
+            if p.id != "openid"
+        ]
+
+        if (
+            openid_provider := next(filter(lambda p: p.id == "openid", providers), None)
+            is not None
+        ):
+            resp += [
+                {
+                    "name": b.name,
+                    "login_url": openid_provider.get_login_url(
+                        request,
+                        process="connect",
+                        openid=b.openid_url,
+                    ),
+                }
+                for b in openid_provider.get_brands()
+            ]
```
Whats the purpose of this special case? And please add coverage for it, this is the only remaining uncovered line

mpflanzer replied:
It's the equivalent to what allauth is doing in their template: https://github.com/pennersr/django-allauth/blob/main/allauth/templates/socialaccount/snippets/provider_list.html
The `openid` provider is a bit special and configured slightly different compared to the others: https://docs.allauth.org/en/latest/socialaccount/providers/openid.html
I'll see to add a test for those lines as well.

shamoon replied:
I couldn't resist. I also dont think that code was working (hence, tests 👍)

`PAPERLESS_APPS="allauth.socialaccount.providers.openid"`
<img width="653" alt="Screenshot 2024-02-05 at 1 04 37 AM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/bf018805-202c-4de4-b198-bb28075fe01f">

mpflanzer replied:
Ah, I think a pair of parenthesis was missing. :see_no_evil:  Nice, thanks for adding the test

shamoon committed:
Revert "Add comment why openid_connect is installed by default"

This reverts commit d1b1f4fcda389c17875cbaaa032be3821cf373f6.

shamoon committed:
Merge branch 'dev' into pr/5190

shamoon committed:
Fix social_account_providers endpoint with openid provider and add coverage

shamoon commented:
> All modified and coverable lines are covered by tests ✅

😎

michelheusschen commented:
Just tried this new feature with Authelia and so far working without any major issues. Thanks to everyone involved!

I've mainly tested the connecting/disconnecting part where I do have some recommendations:
1. Provide a second confirmation when clicking `Disconnect` in profile settings
2. Disallow disconnecting when no password is set to prevent self lockouts

auer1329 commented:
I just tested the latest version with _authentik_, so far it's working fine with the following configuration:

`PAPERLESS_PROXY_SSL_HEADER='["HTTP_X_FORWARDED_PROTO", "https"]'`
`PAPERLESS_SOCIALACCOUNT_PROVIDERS='{ "openid_connect": { "APPS": [ { "provider_id": "authentik", "name": "authentik", "client_id": "<client-id>", "secret": "<secret>", "settings": { "server_url": "https://auth.domain.tld/application/o/<application-slug>/.well-known/openid-configuration" } } ], "OAUTH_PKCE_ENABLED": "True" } }'`

Is there something specific that should be tested in order to verify full compatibility?

shamoon committed:
Mock OIDC provider instead of including in INSTALLED_APPS

shamoon committed:
Add coverage for squirrely serializer method

shamoon commented:
> I just tested the latest version with _authentik_, so far it's working fine with the following configuration:
> 
> Is there something specific that should be tested in order to verify full compatibility?

That's definitely good to know, thanks! But no, at this point the PR just needs documentation and a code review, hopefully no major hurdles left.

shamoon committed:
Fix bootstrap icons

compgeniuses commented:
Documentation, Always the bigger headache

stumpylog reviewed (dismissed):
Just noticed a couple of things (besides finishing up the documentation).  I do still need to grab the latest and test with Authelia, but it appears from other comments it is working, so that's great to hear.

shamoon committed:
Move adapter tests

shamoon commented on src/paperless/test_adapter.py:
Oops, my bad

shamoon commented on src/documents/tests/test_api_profile.py:
```diff
@@ -103,3 +201,136 @@ def test_profile_not_logged_in(self):
+            2,
+        )
+
+    # @mock.patch(
```
Ah yea thanks

shamoon committed:
Remove test comment

mpflanzer commented:
> Disallow disconnecting when no password is set to prevent self lockouts

For this we would need to expose `has_usable_password` in the profile API. Any concerns regarding that @shamoon @stumpylog? If that's fine it's just a small change

mpflanzer committed:
Adding confirmation before disconnect

shamoon commented:
That will be a modal on top of a modal, which doesnt sound great. I'd implement another way to confirm. Also, please cover any new changes with tests.

shamoon commented:
> That will be a modal on top of a modal, which doesnt sound great.

I'd do something like a little text confirmation below the button. I can also do this if you'd prefer.

mpflanzer commented:
Yes, modal on modal is not great. A two step button would be working. First "Disconnect" and then "Are you sure?" and you need to press again. But if that is not already present somewhere in paperless it might not be a small thing to add.

shamoon commented:
I like that idea too, would probably play with it to see which is best. Yes, this doesnt exist anywhere else at the moment.

mpflanzer commented:
I'm not sure I can picture what you mean with the text confirmation. Feel free to revert my changes and have a go at your idea.

shamoon committed:
Revert "Adding confirmation before disconnect"

This reverts commit 732c123d44d193509207528aef88eb5697cfd844.

shamoon committed:
Update frontend strings

shamoon commented:
I've reverted that change for now (also pushed a change to update the frontend strings).

This confirmation thing is not worth holding up this PR, there is already a warning beneath the button and that is how all other things like that work in the app (regenerating API key, for example). I will see about implementing something like that in a separate PR, but lets stay focused to try and get this one over the line.

shamoon committed:
Update adapter.py

shamoon committed:
Disable disconnecting if no password is set, show warning

shamoon commented:
I added a little warning tooltip & disabled the disconnect button if there's no password set 

<img width="298" alt="Screenshot 2024-02-06 at 3 23 03 PM" src="https://github.com/paperless-ngx/paperless-ngx/assets/4887959/2846757a-c6ea-4d9c-a91a-57775c160016">
<br/><br/><br/>

- [ ] Docs
- [ ] Backend review
- [ ] 🚀

mpflanzer commented on src-ui/src/app/components/common/profile-edit-dialog/profile-edit-dialog.component.html:
```diff
@@ -54,7 +54,18 @@ <h4 class="modal-title" id="modal-basic-title" i18n>Edit Profile</h4>
+                <button
+                  type="button"
+                  class="btn btn-outline-danger btn-sm ms-2 align-baseline"
+                  [disabled]="!hasUsablePassword"
```
We could just disable the button if only one social account is remaining and no password is set. Do we also want to prevent the disconnect on the backend in that case?

```
[disabled]="!hasUsablePassword && socialAccounts.length == 1"
```

shamoon replied:
Good idea.

I dont think so, this is just a helpful warning via the UI.

mpflanzer replied:
Ok, thanks for the quick change with the length

shamoon committed:
Also require one remaining social account to disable disconnect

mpflanzer committed:
Add localized strings for the backend

shamoon committed:
Add docs

shamoon committed:
Merge branch 'dev' into pr/5190

shamoon commented:
Again, not clear on the delay but in the interest of moving things along, I've added a first draft of the docs.

shamoon committed:
Fix documentation of default values for a couple of configs

shamoon committed:
Dont require coverage of placeholder populateuser

mpflanzer committed:
Minor additions to the documentation

shamoon committed:
Add PAPERLESS_SOCIALACCOUNT_ALLOW_SIGNUPS

shamoon committed:
Merge branch 'dev' into mpflanzer-allauth

compgeniuses commented:
hurrey, Thumbs up and good work everyone.

I am curious, Can this new implementation be able to work with Ldap authentication concurrently, from the documented Ldap inputs from the discussions

shamoon commented:
Please refer to the [allauth documentation](https://github.com/search?q=repo%3Apennersr%2Fdjango-allauth+ldap&type=code). It looks like not.

mpflanzer commented:
Nice! :tada:  Thanks a lot to everyone for the support and bringing it over the finish line. :pray:

TheRealGramdalf commented:
In the interest of staying not straying off topic I'll keep it short - tl;dr is I was dealing with nearly losing half the data on my computer, so I wasn't really able to come through with the docs. My apologies on that one, sorry if this held things up.

shamoon commented:
Did you try reading the docs? Please start a disucssion if you need help, instead of commenting on the closed PR

ytzelf commented:
Maybe loosely related but now that we can have OIDC auth is there a simple way to disable the legacy login?
Thanks a lot

compgeniuses commented:
@ytzelf  i dont think its recommened of any application, a backup admin account is necessary in software administration

ytzelf commented:
Exactly I would welcome the option to get rid of the legacy user login option. Django admin UI should still be avaialble through /admin. That's how vaultwarden does it if memory serves.

compgeniuses commented:
Similar to how nextcllod does it, once you configure SSO, you can impelment its as the default way to login, however you have the option of logging into the account using the builtin user through a defined URL

florianschroen commented:
I guess it's important to additional disable password (legacy login) for the API (like nextcloud does)

Apps should redirect the User to oidc login in Browser for the first login to get a valid session and generate an Individual access token for further API access.

Atm I need to exclude the API url from the oidc handling because they need the password login for API access. It would be great when this can be "fixed", because then the whole service access can be secured by long tokens or password and MFA.

compgeniuses commented:
@florianschroen  since this matter has already been merged nd closed, why dont we have another issue created to discuss these varied parameters, that way it would be easy if someone wants to make a PR to refence the matter

rdoering commented:
@mpflanzer 
I tried this feature and configured the environment var SOCIALACCOUNT_PROVIDERS with 2.5.3 via Docker , but without success.
Is there a way to get verbose messages, why a access were denied or what went wrong while fetching oidc data?

shamoon commented:
Please dont comment on a closed PR for help, you can start a discussion.
